matter_server/server/device_controller.py

```diff
@@ -274,21 +272,28 @@ async def commission_with_code(
                 attempts,
                 MAX_COMMISSION_RETRIES,
             )
-            result: (
-                PyChipError | None
-            ) = await self._chip_device_controller.commission_with_code(
-                node_id,
-                code,
-                DiscoveryType.DISCOVERY_NETWORK_ONLY
-                if network_only
-                else DiscoveryType.DISCOVERY_ALL,
-            )
-            if result and result.is_success:
-                break
-            if attempts >= MAX_COMMISSION_RETRIES:
-                raise NodeCommissionFailed(
-                    f"Commission with code failed for node {node_id}."
+            try:
+                commissioned_node_id: int = (
+                    await self._chip_device_controller.commission_with_code(
+                        node_id,
+                        code,
+                        DiscoveryType.DISCOVERY_NETWORK_ONLY
+                        if network_only
+                        else DiscoveryType.DISCOVERY_ALL,
+                    )
                 )
+                # We use SDK default behavior which always uses the commissioning Node ID in the
+                # generated NOC. So this should be the same really.
+                LOGGER.info(
+                    "Commissioned Node ID: %s vs %s", commissioned_node_id, node_id
+                )
+                assert commissioned_node_id == node_id
+                break
+            except ChipStackError as err:
+                if attempts >= MAX_COMMISSION_RETRIES:
```

Comment:
Do we still need the retries ?

Reply:
Yeah good question. I did many many commissionings here, in most cases the retry didn't help as the issue is something else (like incorrect code/device no longer in commissioning mode/device already commissioned with the current fabric).

However, I had one case where it indeed helped:

```
2024-06-14 12:27:18.577 (MainThread) VERBOSE [matter_server.server.client_handler] [140550928449168] Handling command commission_with_code
2024-06-14 12:27:18.578 (MainThread) INFO [matter_server.server.device_controller] Starting Matter commissioning with code using Node ID 124 (attempt 1/3).
...
2024-06-14 12:27:53.194 (Dummy-2) CHIP_PROGRESS [chip.native.DMG] Time out! failed to receive invoke command response from Exchange: 37724i
2024-06-14 12:27:53.194 (Dummy-2) CHIP_PROGRESS [chip.native.CTL] Device failed to receive the operational certificate Response: src/app/CommandSender.cpp:328: CHIP Error 0x00000032: Timeout
2024-06-14 12:27:53.194 (Dummy-2) CHIP_ERROR [chip.native.CTL] Error on commissioning step 'SendNOC': 'src/app/CommandSender.cpp:328: CHIP Error 0x00000032: Timeout'
2024-06-14 12:27:53.194 (Dummy-2) CHIP_PROGRESS [chip.native.CTL] Going from commissioning step 'SendNOC' with lastErr = 'src/app/CommandSender.cpp:328: CHIP Error 0x00000032: Timeout' -> 'Cleanup'
2024-06-14 12:27:53.195 (Dummy-2) CHIP_PROGRESS [chip.native.CTL] Performing next commissioning step 'Cleanup' with completion status = 'src/app/CommandSender.cpp:328: CHIP Error 0x00000032: Timeout'
...
2024-06-14 12:27:58.272 (MainThread) INFO [matter_server.server.device_controller] Starting Matter commissioning with code using Node ID 124 (attempt 2/3).
...
2024-06-14 12:28:10.264 (MainThread) INFO [matter_server.server.device_controller] Matter commissioning of Node ID 124 successful.
```

This was a Meross Smart Plug :man_shrugging: 

But in all other cases, it just took longer for the failure to eventually show up, which kinda hurts UX :cry: 

So I think we rather do a single one, and make sure to propage the error to the UI nicely. Maybe we can have a retry button there? :thinking:

Reply:
I've removed the commissioning retries. I think this is indeed more often not helpful. Let's see how this works in practice.